Changes as discussed June 5

wmamills opened the pull request:
New PDF: https://github.com/wmamills/virtio-msg-spec/releases/download/wam-review-06-09-wmamills-draft-v16/wam-review-06-09-wmamills-draft-v16.pdf

V1:
* Eliminate EVENT_CONFIG_EMPTY
* Fix all Virtqueue constants to 0xFFFFFFFF
* Replace DEVICE_ADDED and DEVICE_REMOTE with EVENT_DEVICE
* Add SHM and admin virtqueue support
* Make field order of {SET,GET,EVENT}_CONFIG consistant
* various typos, spelling, and wording fix-ups

V2:
* update github to use unique tags and pdf names
* Review fixup for EVENT_CONFIG[_EMPTY]
* Review fixup for EVENT_DEVICE
* Add cover letter for RFC2

V3:
* dropped github action commit as it came via a different PR
* Reordered cover letter to most recent (and delete mention of "all virtqueues" value)
* Add new cover letter commit to mention FF-A and other upcoming work

wmamills committed:
virtio-msg: Remove EVENT_CONFIG_EMPTY

As decided on June 5th, the EVENT_CONFIG_EMPTY message is deleted.
Instead of documenting "invalid" settings of device status and
configuration generation count, I have removed that use case of
EVENT_CONFIG.

EVENT_CONFIG can still be used with an without changed configuration data
in the payload.  More explicit documentation on these two cases has been
added.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: Fix virtqueue index value for "all virtqueues"

The hex value for this case was always intended to be "all 1s" but was
mis-specified before.  This fixes that.

This also makes it clear that for the EVENT_AVAIL use case, this would
typically be used on the device side.  (The driver side always knows which
virtqueue was updated as it originates this event.)

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: fix spelling and English usage

* writeing -> writing
* support to send -> support sending
* /msgref -> \msgref

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: Use EVENT_DEVICE to replace DEVICE_ADDED and DEVICE_REMOVED

This uses the EVENT_ name for event messages and replaces two messages
with one that includes a state value.  This allows future expansion of
state to handle things like suspend/resume and request removal.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: fix offset of max num virtqueues in GET_DEVICE_INFO

Fix the copy and past error here.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: add support for admin virtqueues and SHM regions

* Add support for device owned shared memory segments from virtio v1.3
* Add support for Admin virtqueues from virtio v1.4 (draft)

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: Make {SET,GET,EVENT}_CONFIG field order more consistent

Make the order of fields:
* [config gen count, if present]
* config offset
* config length
* config data

Signed-off-by: Bill Mills <bill.mills@linaro.org>

bertrand-marquis commented on transport-msg.tex:
```diff
@@ -1037,10 +1051,10 @@ \subsubsection{Overview}
-using out-of-band notifications, using polling, or when emulating or bridging
-to another virtio transport.
+If virtqueue index is 0xFFFFFFFF then it indicates that all virtqueues should be
+polled for available buffers. This may be useful on the device side when the
```
Shouldn't this be "on the driver side" instead of the bus side. If out of band or polling is used this message is generated by the bus with data.

bertrand-marquis replied:
in fact to be coherent with the event used text, it might be better to just no say which side (see sentence at 1077).

wmamills replied:
EVENT_AVAIL is driver -> device.  If the message is not sent in-band then the device side might not know which virtqueue it is for.  In this case the device side BUS can use this value to cause the device side transport to poll each virtqueue of the device.  So the usefulness is on the device side.

The device side bus might not know which device either.  I am assuming that the bus will make a loop per active device and send this event for each.  We COULD do this for virtqueues also but this would mean the bus peeking into the per device transport instance structure.

wmamills replied:
I have deleted this value and description for both EVENT_AVAIL and EVENT_USED, at least for now.  We can always add it later if needed but it may just be better to do the loop in the bus.

Some bus implementations will need to have a loop anyway and the bus having visibility to the common virtio-msg instance structure is not a big deal in most cases.

wmamills replied:
On call June 19, it was agreed to take this out at least for now.

bertrand-marquis reviewed (approved):
Definitely looks good to me.
Just one minor comment

arnopo commented on transport-msg.tex:
```diff
@@ -1019,11 +989,19 @@ \subsubsection{Overview}
 \hline
 \end{tabular}
 
+The number of bytes field may be zero. If so the configuration data field
```
suggestion:

```suggestion
The number of bytes field \emph{MAY} be zero. If so the configuration data field.
```
 Same for the field above

wmamills replied:
OK

Note: Actually the upstream spec only uses MAY not \emph{MAY}

arnopo commented on transport-msg.tex:
```diff
@@ -1015,10 +1015,10 @@ \subsubsection{Overview}
+If virtqueue index is 0xFFFFFFFF then it indicates that all virtqueues should be
+polled for available buffers. This may be useful on the device side when the
+bus implementation is using out-of-band notifications, using polling, or when
+emulating or bridging to another virtio transport.
```
I 'm trying to figure out how it could be used. 
Except if I missed something the 0xFFFFFFFF can only come from the bus, right? the driver give a virtqueue index.
Seems to me that the "may be" could be challenged, requesting a concrete example.

wmamills replied:
See above

arnopo commented on transport-msg.tex:
```diff
@@ -358,14 +357,14 @@ \subsubsection{Device Hotplug}
 \begin{itemize}
-  \item \busref{DEVICE_ADDED} for a newly available device,
-  \item \busref{DEVICE_REMOVED} for a device that is no longer accessible.
+  \item \busref{EVENT_DEVICE} with state ready for a newly available device,
```
suggestion only
```suggestion
  \item \busref{EVENT_DEVICE} with state DEVICE_BUS_STATE_READY indicates that a new device is set up and ready
```

arnopo commented on transport-msg.tex:
```diff
@@ -358,14 +357,14 @@ \subsubsection{Device Hotplug}
-  \item \busref{DEVICE_ADDED} for a newly available device,
-  \item \busref{DEVICE_REMOVED} for a device that is no longer accessible.
+  \item \busref{EVENT_DEVICE} with state ready for a newly available device,
+  \item \busref{EVENT_DEVICE} with state removed for a device that is no longer accessible.
```
```suggestion
  \item \busref{EVENT_DEVICE} with state DEVICE_BUS_STATE_REMOVED indicates a device that is no longer accessible.
```

arnopo commented on transport-msg.tex:
```diff
@@ -1124,21 +1121,29 @@ \subsection{Bus Messages}\label{sec:Virtio Transport Options / Virtio Over Messa
+\hline
+DEVICE_BUS_STATE_READY      & 0x0  & Device is present and ready \\
+\hline
+DEVICE_BUS_STATE_REMOVED    & 0x1  & Device is no longer present \\
```
What about inversing the logic (1 for ready, 0 for removed) or using 0x1 for READED and 0x2 for REMOVED ?

bertrand-marquis replied:
we should not use the value 0 and keep it reserved so using 1 and 2 sounds good

wmamills replied:
There is really only one normal/ready state so that should be first.  There are many types of not ready but we will start with just one "removed".   I will use 1 for ready and 2 for removed.

arnopo commented on transport-msg.tex:
```diff
@@ -468,12 +468,21 @@ \subsubsection{Device Information}
+    shared memory segments this device has.
+  \item \textbf{Admin virtqueue starting index}: The virtqueue index for the
+    first admin virtqueue supported by the device.
+  \item \textbf{Admin virtqueue count}: The number of admin virtqueues
```
Do we need a field for the split/packed queue type?

wmamills replied:
Split/packed queues are negociated via feature bits.  I don't see anything in the virtio-mmio transport registers for split/packed.  If you have a reference please let me know

arnopo replied:
my bad,  `VIRTIO_F_RING_PACKED` do the job at driver/device level

wmamills committed:
virtio-msg: review fixup for EVENT_DEVICE

* Use full state names in description
* Use 1 & 2 as state values

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: review fixup for EVENT_CONFIG

* Use "\emph{MAY} be zero" for config data length

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: EVENT_{AVAIL.USED} delete "all virtqueues" provision

This is hard to describe and real implementations may just use a loop
anyway.  So delete this for now anyway.

A loop is required in the bus when:
* One OoB notification can trigger multiple virtio devices
* One OoB notification can trigger SOME virtqueues but not all

So this value only covers one use case and it may cause more confusion
than value.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: adjust cover letter for RFC2

Document what we have done since RFC1.

Signed-off-by: Bill Mills <bill.mills@linaro.org>

wmamills committed:
virtio-msg: cover letter, mention more todo items

* As discussed Jun 19, mention work on FF-A spec is underway but don't
  promise a date.
* Do likewise for the AMP data structures and first bus message set
* Do likewise for aligned implementations and PATCH v1

Signed-off-by: Bill Mills <bill.mills@linaro.org>
